File: backend/app/routers/claim.py
"""Claim flow endpoints. See docs/API_CONTRACTS.md "Claim flow (/claim)"."""
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies.auth import CurrentUser, get_current_user, require_admin
from app.dependencies.db import get_db
from app.dependencies.pagination import Pagination, pagination_params
from app.schemas.claim import (
    ClaimApproveRequest,
    ClaimCreate,
    ClaimListResponse,
    ClaimOut,
    ClaimRejectRequest,
    ClaimStatus,
)
from app.services import claim_service

router = APIRouter(prefix="/claim", tags=["claim"])


@router.post("", response_model=ClaimOut, status_code=status.HTTP_201_CREATED)
async def submit_claim(
    body: ClaimCreate,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
) -> ClaimOut:
    claim = await claim_service.create_claim(db, current_user, body)
    return claim_service.to_claim_out(claim)


@router.get("", response_model=ClaimListResponse)
async def list_claims(
    status_filter: ClaimStatus = Query(default="pending_review", alias="status"),
    pagination: Pagination = Depends(pagination_params),
    db: AsyncSession = Depends(get_db),
    _admin: CurrentUser = Depends(require_admin),
) -> ClaimListResponse:
    """Auth: admin only. `status` defaults to `pending_review`; newest first."""
    return await claim_service.list_claims(db, pagination, status_filter)


@router.get("/{claim_id}", response_model=ClaimOut)
async def get_claim(
    claim_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(get_current_user),
) -> ClaimOut:
    claim = await claim_service.get_claim(db, claim_id, current_user)
    return claim_service.to_claim_out(claim)


@router.post("/{claim_id}/approve", response_model=ClaimOut)
async def approve_claim(
    claim_id: int,
    body: ClaimApproveRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_admin),
) -> ClaimOut:
    claim, owner_group_granted = await claim_service.approve_claim(
        db, claim_id, current_user.cognito_sub, body.reviewer_notes
    )
    out = claim_service.to_claim_out(claim)
    out.owner_group_granted = owner_group_granted
    return out


@router.post("/{claim_id}/reject", response_model=ClaimOut)
async def reject_claim(
    claim_id: int,
    body: ClaimRejectRequest,
    db: AsyncSession = Depends(get_db),
    current_user: CurrentUser = Depends(require_admin),
) -> ClaimOut:
    claim = await claim_service.reject_claim(db, claim_id, current_user.cognito_sub, body.reviewer_notes)
    return claim_service.to_claim_out(claim)

File: frontend/src/lib/api/claim.ts
// Typed client for /claim — docs/API_CONTRACTS.md "Claim flow (`/claim`)".
import { apiFetch, toQueryString } from "./client";
import type { PaginatedResponse, PaginationParams } from "@/types/common";
import type {
  ApproveClaimInput,
  ClaimQueueItem,
  ClaimResponse,
  ClaimStatus,
  CreateClaimInput,
  RejectClaimInput,
} from "@/types/claim";

/** POST /claim — auth: any authenticated Cognito user (the claimant). */
export async function submitClaim(
  input: CreateClaimInput,
  accessToken: string
): Promise<ClaimResponse> {
  return apiFetch<ClaimResponse>(
    "/claim",
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/**
 * GET /claim — auth: admin. Claims queue, newest first; `status` defaults
 * to `pending_review` server-side.
 */
export async function listClaims(
  params: PaginationParams & { status?: ClaimStatus },
  accessToken: string
): Promise<PaginatedResponse<ClaimQueueItem>> {
  const query = toQueryString({
    status: params.status,
    page: params.page,
    page_size: params.page_size,
  });
  return apiFetch<PaginatedResponse<ClaimQueueItem>>(
    `/claim${query}`,
    { method: "GET" },
    { accessToken }
  );
}

/** GET /claim/{id} — auth: the claimant (own claim) or admin (any claim). */
export async function getClaimById(
  id: number,
  accessToken: string
): Promise<ClaimResponse> {
  return apiFetch<ClaimResponse>(
    `/claim/${id}`,
    { method: "GET" },
    { accessToken }
  );
}

/** POST /claim/{id}/approve — auth: admin. */
export async function approveClaim(
  id: number,
  input: ApproveClaimInput,
  accessToken: string
): Promise<ClaimResponse> {
  return apiFetch<ClaimResponse>(
    `/claim/${id}/approve`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}

/** POST /claim/{id}/reject — auth: admin. */
export async function rejectClaim(
  id: number,
  input: RejectClaimInput,
  accessToken: string
): Promise<ClaimResponse> {
  return apiFetch<ClaimResponse>(
    `/claim/${id}/reject`,
    { method: "POST", body: JSON.stringify(input) },
    { accessToken }
  );
}
